src/cache/cacher.rs

```diff
@@ -69,6 +70,42 @@ pub trait Cacher: Send + Sync {
     fn hash_url(&self, url: &str) -> String {
         blake3::hash(url.as_bytes()).to_string()
     }
+
+    #[cfg(any(
+        feature = "compress-cache-results",
+        feature = "encrypted-cache-results",
+        feature = "cec-cache-results"
+    ))]
+
+    /// A helper function that returns either compressed or encryption results
+    /// Feature flags are required  for this to work
+    async fn compressed_encrypted_results(
+        &mut self,
+        url: &str,
+    ) -> Result<Vec<u8>, Report<CacheError>> {
+        let results = self.cached_results(url).await?;
+
+        let mut bytes = serde_json::to_vec(&results).map_err(|_| CacheError::SerializationError)?;
+
+        #[cfg(feature = "compress-cache-results")]
+        use std::io::Write;
+        let mut writer = brotli::CompressorWriter::new(Vec::new(), 4096, 11, 22);
+        writer.write_all(&bytes);
+        bytes = writer.into_inner();
+        #[cfg(feature = "encrypt-cache-results")]
+        use chacha20poly1305::{
+            aead::{Aead, AeadCore, KeyInit, OsRng},
+            ChaCha20Poly1305,
+        };
+
+        let key = ChaCha20Poly1305::generate_key(&mut OsRng);
+        let cipher = ChaCha20Poly1305::new(&key);
+        let nonce = ChaCha20Poly1305::generate_nonce(&mut OsRng); // 96-bits; unique per message
+        bytes = cipher
+            .encrypt(&nonce, bytes.as_ref())
+            .map_err(|_| CacheError::EncryptionError)?;
+        Ok(bytes)
+    }
```

Comment:
```suggestion
    #[cfg(any(
        feature = "compress-cache-results",
        feature = "encrypted-cache-results",
        feature = "cec-cache-results"
    ))]

    /// A helper function that returns either compressed or encryption results
    /// Feature flags are required  for this to work
    async fn compress_encrypt_results(
        &mut self,
        search_results: &str,
        url: &str,
    ) -> Result<Vec<u8>, Report<CacheError>> {
        let mut bytes = search_results.as_bytes();
        
        #[cfg(feature = "compress-cache-results")]
        {
	        use std::io::Write;
	        let mut writer = brotli::CompressorWriter::new(Vec::new(), 4096, 11, 22);
	        writer.write_all(&bytes);
	        bytes = writer.into_inner();
	        Ok(bytes)
        }
        #[cfg(feature = "encrypt-cache-results")]
        {
	        use chacha20poly1305::{
	            aead::{Aead, AeadCore, KeyInit, OsRng},
	            ChaCha20Poly1305,
	        };
	
		let cipher = CIPHER.get_or_init(|| {
		        let key = ChaCha20Poly1305::generate_key(&mut OsRng);
		        ChaCha20Poly1305::new(&key)
		});
		
		let encryption_key = ENCRYPTION_KEY.get_or_init(
		        || ChaCha20Poly1305::generate_nonce(&mut OsRng), // 96-bits; unique per message
		);
	
	        bytes = cipher
	            .encrypt(&encryption_key, bytes.as_ref())
	            .map_err(|_| CacheError::EncryptionError)?;
	        Ok(bytes)
        }
    }
```

I think you quite get the issue properly, like it was not clear. Sorry, I think we should have explained it in more depth, but we will take care of it next time. 

Actually, the issue talks about compressing/encrypting or both of the search results before putting them into the cache. So you will need to implement a function which can compress/encrypt or both of the search results in the trait, and then we can call them in the Redis and memory caching functions as needed. 

As such, the implementation looks good, but I would suggest you to make the following changes: 
- Renaming the function to `compress_encrypt_results` because I think this will better match the work the function is doing.
- Splitting the encrypt and compress functions into two because this will make it easier to implement the third feature. As the third feature, will require you to implement a third function which can then call the other two function directly.
- Moving the code to generate cipher and encryption keys into `oncelock` which allows generating the encryption key and cipher once globally. Otherwise, generating each new encryption key on a new call of this function will cause the same results to have encrypted different keys, which can cause issues while retrieving results. For example, the results will never be decrypted because the currently generated new key will be different from the key with which the old results were encrypted with.

Moreover, for the `oncelock` implementation you will need to add a new crate in the project which I would suggest adding it by running the following command:

```shell
cargo add chacha20 --no-default-features
```